File: apps/ios/ResumeForge/Models/Ats.swift
import Foundation
import SwiftUI

struct AtsSubscores: Codable, Hashable {
    var parseability: Double
    var keywordMatch: Double
    var structure: Double
    var impact: Double
    var formatting: Double

    /// Ordered (label, value) pairs for rendering the five bars.
    var ordered: [(area: AtsArea, value: Double)] {
        [
            (.parseability, parseability),
            (.keywordMatch, keywordMatch),
            (.structure, structure),
            (.impact, impact),
            (.formatting, formatting),
        ]
    }
}

enum AtsArea: String, Codable, CaseIterable, Identifiable {
    case parseability, keywordMatch, structure, impact, formatting
    var id: String { rawValue }
    var label: String {
        switch self {
        case .parseability: return "Parseability"
        case .keywordMatch: return "Keyword Match"
        case .structure: return "Structure"
        case .impact: return "Impact"
        case .formatting: return "Formatting"
        }
    }
}

enum AtsSeverity: String, Codable, Hashable {
    case critical, warning, suggestion

    var color: Color {
        switch self {
        case .critical: return .red
        case .warning: return .orange
        case .suggestion: return .blue
        }
    }

    var systemImage: String {
        switch self {
        case .critical: return "exclamationmark.octagon.fill"
        case .warning: return "exclamationmark.triangle.fill"
        case .suggestion: return "lightbulb.fill"
        }
    }

    var label: String { rawValue.capitalized }
}

struct AtsFix: Codable, Hashable {
    var description: String
    var autoApplyPatch: [ResumePatch]?
}

struct AtsFinding: Codable, Identifiable, Hashable {
    var id: String
    var severity: AtsSeverity
    var area: AtsArea
    var message: String
    var fix: AtsFix?
}

struct AtsScoreResult: Codable, Hashable {
    var overall: Double
    var subscores: AtsSubscores
    var findings: [AtsFinding]
    var version: String
}

extension Double {
    /// Color band for an ATS score (0...100).
    var atsScoreColor: Color {
        switch self {
        case ..<50: return .red
        case ..<75: return .orange
        default: return .green
        }
    }
}

File: packages/shared/src/ats.ts
import { z } from "zod";
import { ResumePatchSchema } from "./patch";

export const AtsSubscoresSchema = z.object({
  parseability: z.number().min(0).max(100),
  keywordMatch: z.number().min(0).max(100),
  structure: z.number().min(0).max(100),
  impact: z.number().min(0).max(100),
  formatting: z.number().min(0).max(100),
});
export type AtsSubscores = z.infer<typeof AtsSubscoresSchema>;

export const AtsFindingSchema = z.object({
  id: z.string(),
  severity: z.enum(["critical", "warning", "suggestion"]),
  area: z.enum(["parseability", "keywordMatch", "structure", "impact", "formatting"]),
  message: z.string(),
  fix: z
    .object({
      description: z.string(),
      autoApplyPatch: z.array(ResumePatchSchema).optional(),
    })
    .optional(),
});
export type AtsFinding = z.infer<typeof AtsFindingSchema>;

export const AtsScoreResultSchema = z.object({
  overall: z.number().min(0).max(100),
  subscores: AtsSubscoresSchema,
  findings: z.array(AtsFindingSchema),
  version: z.string(),
});
export type AtsScoreResult = z.infer<typeof AtsScoreResultSchema>;

export const ATS_SCORER_VERSION = "1.0.0";

/** Weighting of subscores into the overall score. Sums to 1. */
export const ATS_WEIGHTS: Record<keyof AtsSubscores, number> = {
  parseability: 0.3,
  keywordMatch: 0.25,
  structure: 0.15,
  impact: 0.2,
  formatting: 0.1,
};
